libraries/freertos_plus/standard/https/src/iot_https_client.c

```diff
@@ -710,45 +729,70 @@ static int _httpParserOnBodyCallback(http_parser * pHttpParser, const char * pLo
     _httpsResponse_t * pHttpsResponse = (_httpsResponse_t *)(pHttpParser->data);
     pHttpsResponse->parserState = PARSER_STATE_IN_BODY;
 
-    /* Has the user provided a buffer and is it large enough to fit the body? The
-    case of body buffer not being large enough can happen if the body was received
-    in the header buffer and the body buffer can not fit in all the body. */
-    if((pHttpsResponse->pBodyCur != NULL) && (pHttpsResponse->pBodyEnd - pHttpsResponse->pBodyCur > 0))
-    {
-        /* There are two scenarios when we need to copy data around:
-        1. Some or all of the response body may have been received in the header
-        buffer. If that is the case, we copy the response body received in the
-        header buffer to the user provided body buffer.
-        2. When we receive chunked header, the actual body is separated in
-        multiple chunks which are preceeded by length. For example, a chunked
-        body may look like:
-
-        7\r\n
-        Mozilla\r\n 
-        9\r\n
-        Developer\r\n
-        7\r\n
-        Network\r\n
-        0\r\n 
-        \r\n
-
-        In this case, we want the parsed body buffer to contain actual body only
-        (MozillaDeveloperNetwork in the above example).
+    if((pHttpsResponse->bufferProcessingState == PROCESSING_STATE_FILLING_HEADER_BUFFER) && (pHttpsResponse->isAsync))
+    {
+        /* For an asynchronous response, the buffer to store the body will be available after the headers 
+         * are read first. We may receive part of the body in the header buffer. We will want to leave this here
+         * and copy it over when the body buffer is available in the _readReadyCallback().
          */
-        /* If the response body found by the parser (pLoc) is not equal to the
-        current writable location in the body buffer (_httpsResponse->pBodyCur),
-        it indicates that:
-        - Either the data is in the header buffer and needs to be copied into the
-        body buffer.
-        - Or it is a chunked response and the data needs to be moved up in the
-        body buffer. */
-        if( (pHttpsResponse->pBodyCur + length) <= pHttpsResponse->pBodyEnd )
+        if( pHttpsResponse->pBodyInHeaderBuf == NULL )
         {
-            if(pHttpsResponse->pBodyCur != (uint8_t*)pLoc)
+            pHttpsResponse->pBodyInHeaderBuf = ( uint8_t* )( pLoc );
+            pHttpsResponse->pBodyCurInHeaderBuf = pHttpsResponse->pBodyInHeaderBuf;
+        }
+
+        /* If there is a chunk encoded body in the header buffer, we will want to overwrite the chunk headers with the 
+           actual body. This is so that when the application calls IotHttpsClient_ReadResponseBody(), in the 
+           readReadyCallback(), we can pass the body into the body buffer provided right away. */
+        if(pHttpsResponse->pBodyCurInHeaderBuf != (uint8_t*)pLoc)
+        {
+            memcpy(pHttpsResponse->pBodyCurInHeaderBuf, pLoc, length);
```

Comment:
what is value of ```length```? How are we certain that ```pHttpsResponse->pBodyCurInHeaderBuf``` has enough space to copy ```length``` bytes?

Reply:
Because the http-parser is passed a length and a buffer in the execution function. This this case the buffer is the header buffer and the length is the length of the header buffer. length passed to the callback will never exceed the length passed into execution function.  pHttpsResponse->pBodyCurInHeaderBuf points to memory location at or before the pLoc in the same buffer.